x-pack/plugins/enterprise_search/server/lib/enterprise_search_request_handler.ts

```diff
@@ -7,63 +7,90 @@
 import fetch from 'node-fetch';
 import querystring from 'querystring';
 import {
+  RequestHandler,
   RequestHandlerContext,
   KibanaRequest,
   KibanaResponseFactory,
   Logger,
 } from 'src/core/server';
 import { ConfigType } from '../index';
 
-interface IEnterpriseSearchRequestParams<ResponseBody> {
+interface IConstructorDependencies {
   config: ConfigType;
   log: Logger;
+}
+interface IRequestParams<ResponseBody> {
   path: string;
+  params?: string;
   hasValidData?: (body?: ResponseBody) => boolean;
 }
+export interface IEnterpriseSearchRequestHandler {
+  createRequest(requestParams?: object): RequestHandler<unknown, Readonly<{}>, unknown>;
+}
 
 /**
- * This helper function creates a single standard DRY way of handling
+ * This helper lib creates a single standard DRY way of handling
  * Enterprise Search API requests.
  *
  * This handler assumes that it will essentially just proxy the
  * Enterprise Search API request, so the request body and request
  * parameters are simply passed through.
  */
-export function createEnterpriseSearchRequestHandler<ResponseBody>({
-  config,
-  log,
-  path,
-  hasValidData = () => true,
-}: IEnterpriseSearchRequestParams<ResponseBody>) {
-  return async (
-    _context: RequestHandlerContext,
-    request: KibanaRequest<unknown, Readonly<{}>, unknown>,
-    response: KibanaResponseFactory
-  ) => {
-    try {
-      const enterpriseSearchUrl = config.host as string;
-      const params = request.query ? `?${querystring.stringify(request.query)}` : '';
-      const url = `${encodeURI(enterpriseSearchUrl)}${path}${params}`;
+export class EnterpriseSearchRequestHandler {
+  private enterpriseSearchUrl: string;
+  private log: Logger;
 
-      const apiResponse = await fetch(url, {
-        headers: { Authorization: request.headers.authorization as string },
-      });
+  constructor({ config, log }: IConstructorDependencies) {
+    this.log = log;
+    this.enterpriseSearchUrl = config.host as string;
+  }
 
-      const body = await apiResponse.json();
+  createRequest<ResponseBody>({
+    path,
+    params,
+    hasValidData = () => true,
+  }: IRequestParams<ResponseBody>) {
+    return async (
+      _context: RequestHandlerContext,
+      request: KibanaRequest<unknown, Readonly<{}>, unknown>,
+      response: KibanaResponseFactory
+    ) => {
+      try {
+        // Set up API URL
+        params = params ?? (request.query ? `?${querystring.stringify(request.query)}` : '');
```

Comment:
Could params be an object rather than a string? This would let us "merge" params, if some need to be dynamic and some static.
```
const allParams = {
  ...(params && { params }), // base params
  ...(request.query && { request.query }) // override / mix in dynamic params
}
params = (Object.keys(allParams).length > 0 ? `?${querystring.stringify(allParams)}` : '');
```

Reply:
Ah! Great call, thanks!

Reply:
f6a709d - this is so much better, thanks for the amazing suggestion!!

I also realized while testing it that `request.query` is also an empty obj so in reality our `request.query ?` check was never failing 🤦‍♀️